Add padding-free to bamba

garrett361 opened the pull request:
# What does this PR do?

Adds padding-free training to the `BambaModel`, enabling more efficient training with causal masking between disjoint sequences.

**Performance**: approximately 2x throughput improvements over naive padding for supervised finetuning on the [Tulu v3](https://arxiv.org/pdf/2411.15124) dataset with [open-instruct](https://github.com/allenai/open-instruct).  Tokens/sec/gpu plots for `batch_size_per_gpu = 4`:

**8 A100s**: 600 --> 1200 Tok/s/gpu

<img width="287" alt="Scherm­afbeelding 2025-01-16 om 3 52 41 PM" src="https://github.com/user-attachments/assets/f36847ec-3c9a-478e-bcd0-f9e25ba626e6" />

**32 A100s**: 450 --> 750 Tok/s/gpu

<img width="287" alt="Scherm­afbeelding 2025-01-16 om 3 52 33 PM" src="https://github.com/user-attachments/assets/f0f6845d-7190-414f-93f2-e8b7ee6f57b5" />

CC @fabianlim 

CC reviewers of #34982: @ArthurZucker @molbap 

## Notes on Code

* `BambaAttention` layers are untouched; only the `BambaMixer` mamba layer code is altered.
* The padding-free path is only supported on `cuda` and requires the mamba kernels.
* Supports both the `position_ids` and `FlashAttentionKwargs` padding-free code paths. 


## Notes on Tests

On both latest `main` and this PR branch the following `tests/models/bamba/test_modeling_bamba.py` tests are failing (with `RUN_SLOW=1`):
```
BambaModelTest::test_eager_matches_fa2_generate
BambaModelTest::test_flash_attention_2_padding_matches_padding_free_with_position_ids
BambaModelTest::test_sdpa_can_compile_dynamic
BambaModelTest::test_torchscript_output_attentions
BambaModelTest::test_torchscript_output_hidden_state
BambaModelTest::test_torchscript_simple
BambaModelIntegrationTest::test_simple_generate
```
* The `test_eager_matches_fa2_generate` test seems flaky: sometimes it passes, other times it fails.
* For `test_flash_attention_2_padding_matches_padding_free_with_position_ids`:
    * On `main`, this test fails because padding-free is not implemented.
    * On this PR branch this test fails because this PR only uses `position_ids` when `model.training = True` and this test [explicitly calls](https://github.com/huggingface/transformers/blob/72d1a4cd53d90d5db384df948ccc293b3c1e3b9d/tests/test_modeling_common.py?plain=1#L4535) `eval()` on the model.  I have checked that this test passes when `model.training = True`. **Edit**: see `BambaModelTest::test_attn_mask_position_ids_flash_attn_equality`, also.
* `test_simple_generate` appears to just need a simple edit for its expected text. It consistently fails with:
```
AssertionError: '<|be[35 chars]on this lovely evening? I hope you are all doing well. I am' != '<|be[35 chars]on this lovely evening? I hope you are all having a good time.'
```
where the generated and expected text differs at the very end.


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Rocketknight1 commented:
cc @arthurzucker for bamba, but let me know if you want me to take a look since it seems like quite an extensive PR!

garrett361 commented:
> it seems like quite an extensive PR!

I don't think it's very many changes, ultimately! Basically it just adds two helper functions so that `position_ids` and `FlashAttentionKwargs` get properly converted to the `seq_idx` arg that mamba expects:
* `get_cu_seq_lens_from_position_ids`
* `get_seq_idx_from_cu_seq_lens`

So, basically the above, making sure `**kwargs` get passed everywhere they should, and a little code cleanup.

garrett361 commented:
Added a commit with `BambaModelTest::test_attn_mask_position_ids_flash_attn_equality` which tests the various code paths against each other.

garrett361 commented:
Hi @ArthurZucker @molbap, please let me know if I can answer any questions about this PR, thank you!

ArthurZucker commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -940,10 +931,39 @@ def extra_repr(self):
         return f"{tuple(self.weight.shape)}, eps={self.variance_epsilon}"
 
 
+def get_cu_seq_lens_from_position_ids(position_ids: torch.LongTensor) -> torch.LongTensor:
```
not super accurate to use positions ids because they can include padding

garrett361 replied:
Yes, I thought about this for a while and it's why I am checking the `non_increasing_pos_id` condition.

The (just-updated; fixed an error) helper now reads:

```py
def get_cu_seq_lens_from_position_ids(position_ids: torch.LongTensor) -> torch.LongTensor:
    batch_size = position_ids.shape[0]
    if batch_size != 1:
        raise ValueError("Only batch size 1 is supported.")
    device = position_ids.device
    idxs = torch.arange(1, position_ids.shape[1], device=device)
    non_increasing_pos_id = position_ids[0, 1:] <= position_ids[0, :-1]
    next_pos_is_is_zero = position_ids[0, 1:] == 0
    new_seq_idxs = non_increasing_pos_id | next_pos_is_is_zero
    cu_seq_lens = torch.empty(new_seq_idxs.sum() + 2, device=device, dtype=torch.int64)
    cu_seq_lens[0], cu_seq_lens[1:-1], cu_seq_lens[-1] = 0, idxs[new_seq_idxs], position_ids.shape[-1]
    return cu_seq_lens
```

My goal was to treat every padding token (assumed to be encoded as a negative number, like -100) as an individual sequence, while treating the non-padding sequences correcty.  

So, an extreme case like  like `position_ids = [-100, 0, 1, -100, -100, 0, 1, 2]`) would turn into:
```
# get_cu_seq_lens_from_position_ids
cu_seq_lens = [0, 1, 3, 4, 5, 8]
```
thanks to the non-increasing pos id check. Seems reasonable to me, since:
1. The non-trivial sequences are assigned the correct lengths
2. Every padding tok is a new, len 1 segment, so no unnecessary compute will be expended in attending across a span of padding toks.

Compare to what is done in `prepare_fa2_from_positions_ids`, which I was trying to improve upon:

https://github.com/huggingface/transformers/blob/c772bff31a65c9c6002d0e74797cb130959a3716/src/transformers/modeling_flash_attention_utils.py#L174-L179

Passing the same `position_ids` into the above gives:
```
# prepare_fa2_from_positions_ids
cu_seq_lens = [1, 5, 8]
```
which incorrectly implies there's a subseq of len 4, since it only checks for `pos_id = 0`.

Thoughts? Very open to improvements!

ArthurZucker reviewed (commented):
Hey! Good work and good PR, one thing is that for training we do recommend using the padding-free data collator that takes care of the flattening and passing approriate kwargs. This prevents us from having to do too many changes, appart from poping the cu seqlens if they are inputs

garrett361 commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -940,10 +931,39 @@ def extra_repr(self):
+            for idx, n in enumerate(torch.diff(cu_seq_lens, dim=-1))
+        ]
+    )
+    return seq_idx[None]
```
Good call, I minimized the allocations in all three helper functions.

garrett361 commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -998,11 +1018,29 @@ def forward(
+                else:
+                    seq_idx = get_seq_idx_from_cu_seq_lens(cu_seq_lens)
+            else:
+                seq_idx = None
```
Done, much better now.

garrett361 commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -516,28 +516,17 @@ def cuda_kernels_forward(
         cache_params: Optional[HybridMambaAttentionDynamicCache] = None,
-        cache_position: Optional[torch.LongTensor] = None,
         attention_mask: Optional[torch.Tensor] = None,
+        seq_idx: Optional[torch.Tensor] = None,
```
The mamba kernels use `seq_idx` to encode the packing of multiple sequences together, rather than `cu_seq_lens` or `position_ids`.  The three are related like:

```
position_ids = [0, 1, 2, 0, 1, 0, 1, 2, 3, 4, 0, ... ] # position of each token within their seq
cu_seq_lens = [0, 3, 5, 9, 10, ...] # cumulative seq lens across seqs
seq_idx = [0, 0, 0, 1, 1, 2, 2, 2, 2, 3, ...] # idx the sequences directly
```

I'll add this info to the `BambaMixer` doc string.

garrett361 commented:
> Good work and good PR

Thank you!

> one thing is that for training we do recommend using the padding-free data collator that takes care of the flattening and passing approriate kwargs. 

Yep, just trying to support all code paths. Hope I didn't misunderstand you here.

I think I have addressed all comments @ArthurZucker , let me know!

I also had to run `modular_model_convertor.py` on all models to get past CI; some other PR seemed to cause minor CI breakage.

ArthurZucker reviewed (commented):
I am aligned with you on enable the path! 
The key idea for me is that all these utilities, and calls you are doing are coming from not properly preparing the inputs! 
And IMO preparing them outside the forward loop is a lot better! You leave it to the data collator. 

What I am saying here is I'd rather we add a data collator for bamba, that properly takes care of initializing the seq_idx for example, than having to add extra logic. 
BUT we need to propagate the seq_idx for sure! 

Does it make sense for you?

garrett361 commented:
> The key idea for me is that all these utilities, and calls you are doing are coming from not properly preparing the inputs!
And IMO preparing them outside the forward loop is a lot better!

Ah, yes I agree. It would be much better to just have `seq_idx` prepared once at the outset in the dataloader (along with `position_ids` and/or `cu_seq_len_x`), rather than doing the conversions in the model.  Though we still the model to be able to do these conversions if a suboptimal dataloader is used, yes?

Another optimization I could do is to move the conversion code out of the repeated `BambaDecoder` layers and into `BambaModel`, so we only convert once. I put these helper functions inside the `mamba` branch of the `BambaDecoder` layer because I was trying to be conservative and not touch the full-attention layer code path at all, but this isn't ideal for perf.  Thoughts?

Last, if you haven't seen it, some similar conversation is happening over in #35941; see [this comment.](https://github.com/huggingface/transformers/pull/35941#issuecomment-2639515031)

vasqu commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -899,15 +877,38 @@ def forward(
+            and cache_position is not None
+            and cache_position[0] > 0
+        )
+        hidden_states = apply_mask_to_padding_states(hidden_states, attention_mask)
```
A small nit, but idk why this has been moved - I'd rather leave it in the respective forward calls to see the step by step things that go on without mixing modelling logic between the functions too much.

vasqu replied:
Only talking about the very last line marked here, i.e. `hidden_states = apply...`

garrett361 replied:
Just a personal preference refactoring, but I can revert it.

garrett361 replied:
Reverted.

vasqu commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -1208,7 +1265,17 @@ def forward(
+            raise ValueError(
+                "Padding-free training using position_ids or FlashAttentionKwargs requires ",
+                "the flash_attention_2 attention implementation and mamba cuda and triton kernels.",
+            )
```
Idk how I feel about this - on the one hand I understand simplifying things, i.e. we need both, but it also heavily restricts usage imo and not everyone wants to bloat their setups with all libraries (maybe Im the only one :)) and I could run the mamba padding free without needing flash attn.

Also, it feels like we're mixing and interweaving things that come from other stuff which is unideal ~ the naming is / will get confusing imo. Would be nice to have unified interface for padding free or something without being constraint to flash attn only.

garrett361 replied:
> heavily restricts usage imo and not everyone wants to bloat their setups with all libraries

Yeah, it would be nice to support padding-free for all dependency paths, but I just wanted to start with the most performant cases. Happy to follow up on this with a pure torch path impl.  Didn't want to create some massive PR which had both as my first contribution.

> I could run the mamba padding free without needing flash attn
Are you saying this is the case today? Or you wish it were?

> Also, it feels like we're mixing and interweaving things that come from other stuff which is unideal ~ the naming is / will get confusing imo.

Absolutely. It's unfortunate that padding-free is encoded in the FA and mamba kernels in different ways.

vasqu replied:
> Yeah, it would be nice to support padding-free for all dependency paths, but I just wanted to start with the most performant cases. Happy to follow up on this with a pure torch path impl. Didn't want to create some massive PR which had both as my first contribution.
Totally understandable. I think this is up to future PRs but raising concerns early in hope it won't be overlooked/forgotten.

> Are you saying this is the case today? Or you wish it were?
More like I wish it were ^^

vasqu reviewed (commented):
Love the overall idea and implementation. Just my 2 cents on mainly design and maybe how the future would look like using flash attn kwargs.

Imo it would be nice to introduce some more uniform kwargs that is aligned with padding free, not only flash attn (likely something for the future rather than this PR).

garrett361 commented:
> Love the overall idea and implementation.

Thanks!

> Imo it would be nice to introduce some more uniform kwargs that is aligned with padding free, not only flash attn (likely something for the future rather than this PR).

Agreed, that would be very nice.

ArthurZucker reviewed (commented):
Hey! Super sorry to be a bit strict on this:
> Ah, yes I agree. It would be much better to just have seq_idx prepared once at the outset in the dataloader (along with position_ids and/or cu_seq_len_x), rather than doing the conversions in the model. Though we still the model to be able to do these conversions if a suboptimal dataloader is used, yes?

No 😓 because if we do, we open the door to doing the same for flex attention, then flash attention3 then etc and etc! 

Let's rather create a path (meaning support passing them as kwargs typed properly!) and a datacollator + add doc about this 🤗

garrett361 commented:
> Super sorry to be a bit strict on this:

I understand! I actually personally prefer the more-strict approach, I was just trying to change external APIs as minimally as possible. 

Now that I understand the goals better, here is my plan:
1.  Rework this PR so that all padding-free related tensors appear as proper kwargs, raising `ValueError`s if improper combinations are passed in.
2. Open a separate PR which enables `DataCollatorWithFlattening` to optionally produce `seq_idx` and/or the `FlashAttentionKwargs`.
3. Open a similar PR in `trl` for `DataCollatorForCompletionOnlyLM`

How does that sound @ArthurZucker ?

An aside: making all of these padding-free args proper individual kwargs also has the benefit of enabling (padding-free + activation checkpointing) for various models, since the latter feature needs explicit args.  The `FlashAttentionKwargs` path for llama models does not support grad checkpointing: `**flash_attn_kwargs` are passed to the non-grad-ckpt layer, but no the checkpointed one, since checkpointing doesn't support kwargs.

https://github.com/huggingface/transformers/blob/60226c6ff3d6bb225782341e58cce5d31f5be1c7/src/transformers/models/llama/modeling_llama.py?plain=1#L581-L604

garrett361 commented:
@ArthurZucker could you please advise on the above plan? Would love to implement whatever path you think is best, but would like confirmation before doing unnecessary work.

garrett361 commented:
Started a draft PR for adding more return values to the `DataCollatorWithFlattening` here: #36456

garrett361 commented:
I reworked this PR so that it doesn't introduce any tensor conversions for padding-free training.  The user must either provide:
* `seq_idx` and `position_ids`
* `seq_idx`, `position_ids` and all of the `FlashAttentionKwargs` (for better `torch.compile` compatibility).
The model code raises a `ValueError` if improper combinations are provided.  The intention is that padding-free for `bamba` will be used along with `DataCollatorWithFlattenting` with additional return values which will be provided by #36456.  

I made the `forward` signature explicit nearly everywhere(*) and added some doc strings.

(*)The one exception is `BambaForCausalLM` where all of the padding-free args are still processed through `**kwargs`.  Adding `seq_idx, cu_seq_lens_q, ...` to `BambaForCasualLM` in `modular_bamba.py` does not cause `utils/modular_model_coverter.py` to propagate these args non-trivially to the `BambaForCausalLM` class in `modeling_bamba.py`, presumably because `BambaForCasualLM` inherits from `LlamaForCausalLM` which doesn't use these args.  Do we want to make all of these `FlashAttentionKwargs` explicit in `LlamaForCausalLM`, too, eventually?

@ArthurZucker @vasqu please let me know what you think.

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -1202,6 +1283,8 @@ def forward(
         return_dict: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
         logits_to_keep: Union[int, torch.Tensor] = 0,
+        # NOTE: @goon - padding-free kwargs intentionally omitted. Need to change LlamaForCausalLM,
+        # or the code gen tool, for explicitly included padding-free kwargs be properly processed.
         **kwargs,
```
modular cc @Cyrilvallez @ArthurZucker

vasqu commented on tests/models/bamba/test_modeling_bamba.py:
Would be nice to test the error paths with invalid combinations e.g. only `max_seq_q` provided.

garrett361 replied:
added a test along these lines

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -1023,6 +1098,12 @@ def forward(
+                    max_length_q=max_length_q,
+                    max_length_k=max_length_k,
+                    seq_idx=seq_idx,
+                    **kwargs,
```
```suggestion
```
Unnecessary?

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -949,8 +993,32 @@ def forward(
+                "Either all of (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k)"
+                " must be None, or they must all be provided."
+            )
+        if flash_attn_kwargs_all_provided and (position_ids is None or seq_idx is None):
+            raise ValueError(
+                "If (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k) are provided,"
+                " then position_ids and seq_idx must also be provided."
+            )
```
```suggestion
        if flash_attn_kwargs_all_provided and position_ids is None:
            raise ValueError(
                "If (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k) are provided,"
                " then position_ids must also be provided."
            )
```
Could we not have fa kwarg and position ids? This would enable fa path but not the mamba2 path - which shouldn't be an issue imo.

Might be nice to add for which path here.

garrett361 replied:
Ah, I guess that's possible if the `cu_seq_lens` are trivial like `(0, seq_len)`, yeah.

The concern I had here is silent incorrectness if non-trival `cu_seq_lens = (0, seq_len0, seq_len0 + seq_len1, ...)` are passed in without a corresponding non-trivial `seq_idx`, in which case the outputs would be silently incorrect.

I'll sharpen up the checks here, thanks.

vasqu replied:
Argh yea that's painful :/

vasqu replied:
No worries take your time :)

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -949,8 +993,32 @@ def forward(
+                " then position_ids and seq_idx must also be provided."
+            )
+        if seq_idx is not None and position_ids is None:
+            raise ValueError("If seq_idx is provided, position_ids must also be provided.")
```
Here also which path, i.e. mamba2 path

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -668,15 +649,37 @@ def forward(
+        cu_seq_lens = [0, 3, 5, 6, ...] # cumulative seq lens across seqs
+        seq_idx = [0, 0, 0, 1, 1, 2, 3, ...] # idx the sequences directly
+
+        `seq_idx` is only supported on the fast CUDA path, currently.
```
```suggestion
        `seq_idx` is currently only supported for the fast CUDA path.
```

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -668,15 +649,37 @@ def forward(
-            # tune out hidden states for pad tokens, see https://github.com/state-spaces/mamba/issues/66
-            hidden_states = (hidden_states * attention_mask[:, :, None]).to(dtype)
+        """
+        The `seq_idx` arg is used by the mamba2 trion kernels to handle packed sequences, similarly to how
```
`trion <> triton` typo is still there 😄

vasqu reviewed (commented):
I like the overall implementation, just some comments on the checks at the beginning of the forward and possibly add tests that ensure that errors are raised correctly.

Would let the others comment on modular :D would be nice to enable more (kw)args from a given class.

vasqu commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -949,8 +993,32 @@ def forward(
+        seq_idx: Optional[torch.IntTensor] = None,
+        **kwargs,
     ) -> Union[Tuple, BaseModelOutputWithPast]:
+        # Padding-free args correctness checks: seq_idx and position_ids must always be provided
```
Based on the discussions in #36456 it might be appropriate to wait for the fa utils update and then move these checks (for the fa part) imo. Otherwise, we will need to update this again afterwards. Wdyt

garrett361 committed:
add seq_idx and fa kwargs

garrett361 committed:
update tests

garrett361 committed:
docs and grad ckpt support

garrett361 committed:
fmt

garrett361 committed:
better names

garrett361 committed:
test_raise_missing_padding_free_kwarg_errs

vasqu commented:
Nice! I'll try to have a look at it tomorrow 🤗

vasqu reviewed (approved):
LGTM! Thx a lot for sticking through those long PRs :heart: 

I only think it would be nice to add the performance gains in the documentation of Bamba, e.g. under `Padding Free Training` or something similar. In general the docs may need a revision in regards to perf optimizations (out of scope of this PR :D)

vasqu commented on tests/models/bamba/test_modeling_bamba.py:
```diff
@@ -478,6 +488,139 @@ def _prepare_model_kwargs(input_ids, attention_mask, signature):
+                    for i, a in zip(inputs_dict["input_ids"], inputs_dict["attention_mask"])
+                ]
+
+                # add position_ids + fa_kwargs
```
Same below

garrett361 replied:
np!

garrett361 commented on setup.py:
```diff
@@ -466,7 +466,12 @@ def run(self):
     zip_safe=False,
     extras_require=extras,
-    entry_points={"console_scripts": ["transformers=transformers.commands.transformers_cli:main", "transformers-cli=transformers.commands.transformers_cli:main_cli"]},
+    entry_points={
+        "console_scripts": [
+            "transformers=transformers.commands.transformers_cli:main",
+            "transformers-cli=transformers.commands.transformers_cli:main_cli",
+        ]
+    },
```
Yeah, `make style` or similar forced it, oddly.

garrett361 committed:
+ seq_idx in doc strings

garrett361 committed:
padding free training docs

vasqu commented on docs/source/en/model_doc/bamba.md:
```diff
@@ -63,7 +63,34 @@ response = model.generate(**inputs, max_new_tokens=64)
+the examples are of varying lengths, padding-free training can provide significant speed ups and
+memory savings compared to batching the examples together and using padding, as the unnecessary
+compute and memory due to padding is avoided entirely. The performance gains depend on factors such
+as the model and the data distribution, but throughput gains up to ~2x are commonly seen.
```
Let's add the images of the PR here. Numbers are nice, but graphs are nicer imo.

garrett361 replied:
What about a link to the PR? The images are not the best. I'm also not finding similar plots in other docs (maybe my search skills are failing) and am hesitant to do anything too out of the box.

vasqu replied:
Yea, makes sense. I'd leave the decision to a core maintainer and see what they think :eyes: cc @ArthurZucker @Cyrilvallez

garrett361 committed:
add link to pr plots

garrett361 committed:
raise err on attn_mask with padding free

garrett361 commented:
Reminder @Cyrilvallez @ArthurZucker , thank you.

Also CC @vasqu

ArthurZucker commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -1258,6 +1312,11 @@ def forward(
                     use_cache=use_cache,
                     cache_position=cache_position,
                     position_embeddings=position_embeddings,
+                    cu_seq_lens_q=cu_seq_lens_q,
+                    cu_seq_lens_k=cu_seq_lens_k,
+                    max_length_q=max_length_q,
+                    max_length_k=max_length_k,
+                    seq_idx=seq_idx,
```
why don't we just use kwargs everywhere?

garrett361 replied:
Do you mean pack all of these `forward` method padding free arguments into ` BambaModel.forward(..., *kwargs)` and just call 
```py
decoder_layer(
    hidden_states, 
    ...,
    **kwargs
)
```
here?

> Let's rather create a path (meaning support passing them as kwargs typed properly!)

I thought above you were asking me to enumerate all of these padding free args as individually type keyword arguments.

ArthurZucker commented on src/transformers/models/bamba/modular_bamba.py:
```diff
@@ -965,6 +998,22 @@ def forward(
         if (input_ids is None) ^ (inputs_embeds is not None):
             raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
 
+        padding_free_kwargs = (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k, seq_idx)
+        num_padding_free_kwargs_used = sum(k is not None for k in padding_free_kwargs)
+        if num_padding_free_kwargs_used:
+            if num_padding_free_kwargs_used != len(padding_free_kwargs):
+                raise ValueError(
+                    "All of (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k, seq_idx) must be specified for padding-free training."
+                )
+            if position_ids is None:
+                raise ValueError(
+                    "position_ids must also be specified when (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k, seq_idx) are provided for padding-free training."
+                )
+            if attention_mask is not None:
+                raise ValueError(
+                    "attention_mask must be None when providing (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k, seq_idx, position_ids) for padding-free training."
+                )
```
in general this should not be exposed here !

garrett361 replied:
Where should sanity checks be placed? Or should I just remove them?

ArthurZucker reviewed (commented):
Super happy to have padding free, but it should follow what we have for other model: typed kwargs. This should still work with ckping (partial) and should not need much new code!

garrett361 commented:
> Super happy to have padding free, but it should follow what we have for other model: typed kwargs

@ArthurZucker I guess I misunderstood the ask here. By typed kwargs I thought you meant explicitly adding the padding-free kwargs to the `forward` method with type annotations, like
```py
        cu_seq_lens_q: Optional[torch.LongTensor] = None,
        cu_seq_lens_k: Optional[torch.LongTensor] = None,
        max_length_q: Optional[int] = None,
        max_length_k: Optional[int] = None,
        seq_idx: Optional[torch.IntTensor] = None,
```
What did you mean instead?

garrett361 commented:
> This should still work with ckping (partial) and should not need much new code!

So this would work?

```py
    def forward(
        self,
        input_ids: Optional[torch.LongTensor] = None,
        ...,
        **kwargs,  
    ) -> BaseModelOutputWithPast:
        ...
        if self.gradient_checkpointing and self.training:
                layer_outputs = self._gradient_checkpointing_func(
                    decoder_layer.__call__,
                    hidden_states,
                    ...,
                    **kwargs
                )
```

For some reason, I thought `**kwargs` weren't supported in grad ckpting.

garrett361 committed:
rm raising missing padding free err test

garrett361 committed:
BambaFlashAttentionKwargs

garrett361 commented:
@ArthurZucker I reworked this to use `**kwargs: Unpack[BambaFlashAttentionKwargs]`, with the new class defined to add `seq_idx` to the list of kwargs usually in `FlashAttentionKwargs`.   I also removed the arg checking code. Let me know if I'm still misunderstanding the ask, here, thanks!

ArthurZucker commented on src/transformers/models/bamba/modeling_bamba.py:
```diff
@@ -863,9 +891,15 @@ def forward(
         cache_position: Optional[torch.LongTensor] = None,
         attention_mask: Optional[torch.Tensor] = None,
+        seq_idx: Optional[torch.IntTensor] = None,
+        **kwargs,
```
prob unpack here

ArthurZucker reviewed (approved):
Okay now LGTM! can you just fix the red ci and we can merge!

garrett361 commented:
Thanks so much @ArthurZucker for your patience!

garrett361 committed:
run modular util for modular_granitemoehybrid.py

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35861). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

garrett361 commented:
@ArthurZucker CI is green now, CI just building docs

garrett361 commented:
I probably should have understood the ask better from context, no worries!

vasqu commented:
Very nice contribution, thanks @garrett361:)

garrett361 commented:
Thanks for your help and patience also @vasqu !
